### PR TITLE
[DC-779]Add UI code to call the service to get concept lists with counts

### PR DESCRIPTION
**Pull request by rjohanek**
### What
* in ajax/DatasetBuilder and ajax/DataRepo implement the call from the UI to the getConcepts endpoint /api/repository/v1/datasets/{id}/snapshotBuilder/concepts/{conceptId}
* passing around either the dataset object or just the datasetId which is needed for the ajax call
    * in dataset-builder/DatasetBuilder pass the entire dataset object to ConceptSetCreator since the datasetId is needed and the component is already using the settings which also comes from the dataset object. Also pass in just the datasetId to DomainCriteriaSelector since it is all it needs from the dataset object.
    * in ConceptSetCreator pass just the datasetId to ConceptSelector 
    * in DomainCriteriaSelector pass in just the datasetId to ConceptSelector

### Why
- Users will want to choose from real Domain Concepts that exist in the dataset at hand.

### Testing strategy
- [ ] Create a new cohort on the snapshot builder page. Verify that the concepts that populate for domains are different from what populates on dev. The Concepts on dev represent mocked values in the UI, while the values that populate on this branch represent Concepts stubbed in jade-data-repo. Verify that these values are being successfully retrieved from the endpoint: /api/repository/v1/datasets/{id}/snapshotBuilder/concepts/{conceptId}

### CODE REVIEW

File: src/dataset-builder/TestConstants.ts [hunk 1]
```diff
@@ -100,3 +100,31 @@ export const dummyDatasetDetails = (datasetId: string): DatasetModel => ({
...
+  { id: 203, name: 'Heart Surgery', count: 100, hasChildren: false },
+  { id: 204, name: 'Cancer Surgery', count: 100, hasChildren: false },
+
+  { id: 300, name: 'Observation', count: 100, hasChildren: true },
+  { id: 301, name: 'Blood Pressure', count: 100, hasChildren: false },
+  { id: 302, name: 'Weight', count: 100, hasChildren: false },
+  { id: 303, name: 'Height', count: 100, hasChildren: false },
+];
+
+export const getMockConceptForId = (id: number): Concept => {
```
Comment (rjohanek) [inline src/dataset-builder/TestConstants.ts]: Renamed as a reminder of mocked functionality. Particularly because it still has one use outside of testing that will need to be updated in another PR

File: src/dataset-builder/ConceptSelector.ts [hunk 2]
```diff
@@ -65,13 +67,14 @@ export const ConceptSelector = (props: ConceptSelectorProps) => {
...
         getChildren,
+        datasetId,
       }),
     ]),
     cart.length !== 0 &&
       h(ActionBar, {
         prompt: cart.length === 1 ? '1 concept selected' : `${cart.length} concepts selected`,
         actionText,
-        onClick: () => _.flow(_.map(getConceptForId), onCommit)(cart),
+        onClick: () => _.flow(_.map(getMockConceptForId), onCommit)(cart),
```
Comment (rjohanek) [inline src/dataset-builder/ConceptSelector.ts]: The cart is still mocking the a get concept by id call. I believe we will want to replace this with a call to a real Concept endpoint getById. Is that right? I added to mock to the name to remind us this is not actually getting the concept by id, but I can add a to do and make a new ticket as well if it doesn't already exist.

File: src/components/TreeGrid.ts [hunk 3]
```diff
@@ -52,7 +52,9 @@ type TreeGridProps<T extends RowContents> = {
   /** the initial rows to display */
   readonly initialRows: T[];
   /** Given a row, return its children. This is only called if row.hasChildren is true. */
-  readonly getChildren: (row: T) => Promise<T[]>;
+  readonly getChildren: (datasetId: string, row: T) => Promise<T[]>;
```
Comment (pshapiro4broad) [inline src/components/TreeGrid.ts]: I don't think this is correct, I'll have to take a closer look but there should be a way to use `datasetId` in the dataset builder code without having to add it to the `TreeGrid` API.

Comment (rjohanek) [reply src/components/TreeGrid.ts]: I added it to TreeGrid props because getChildren is called in TreeGridInner which gets its props from TreeGrid. But maybe theres another way to ensure TreeGridInner has the datasetId when it needs it

File: src/dataset-builder/TestConstants.ts (hunk 1 above)
Comment (pshapiro4broad) [reply src/dataset-builder/TestConstants.ts]: This is OK but I would put `mock` at the beginning or end of this name instead of the middle, either `mockGetConceptForId()` or `getConceptForIdMock()`.

Comment (pshapiro4broad) [reply src/dataset-builder/TestConstants.ts]: Although strictly speaking this isn't a mock but a dummy or test version of the API. That is, it isn't created using the jest mocks API.

**Review by nmalfroy — APPROVED** [review]
lgtm!

File: src/dataset-builder/ConceptSelector.ts [hunk 4]
```diff
@@ -71,7 +72,7 @@ export const ConceptSelector = (props: ConceptSelectorProps) => {
       h(ActionBar, {
         prompt: cart.length === 1 ? '1 concept selected' : `${cart.length} concepts selected`,
         actionText,
-        onClick: () => _.flow(_.map(getConceptForId), onCommit)(cart),
+        onClick: () => _.flow(_.map(mockGetConceptForId), onCommit)(cart),
```
Comment (pshapiro4broad) [inline src/dataset-builder/ConceptSelector.ts]: It seems incorrect to me to call a mock directly from non-test code. If the API needs to be mocked in production code because it's not yet implemented, it should be done in the ajax layer, not in the UI layer.

But maybe the fix here is to store `Concept` in `cart` instead of `concept.id`. If we have a `Concept` already it seems like we should store it in the cart instead of its ID.

File: src/dataset-builder/ConceptSetCreator.test.ts [hunk 5]
```diff
@@ -1,22 +1,22 @@
...
-import { SnapshotBuilderSettings } from 'src/libs/ajax/DataRepo';
 import { renderWithAppContexts as render } from 'src/testing/test-utils';
 
 import { ConceptSetCreator, toConceptSet } from './ConceptSetCreator';
 import { homepageState } from './dataset-builder-types';
 import { dummyDatasetDetails } from './TestConstants';
 
 describe('ConceptSetCreator', () => {
-  const datasetDetails = dummyDatasetDetails('0');
+  const datasetId = '0';
```
Comment (pshapiro4broad) [inline src/dataset-builder/ConceptSetCreator.test.ts]: It looks like `datsetId` is unused here except on line 12, is there a reason to declare it separately vs using it inline as before? Also I wonder if `dummyDatasetDetails` needs to accept an `id` parameter in the first place, is there a need to have more than one details object with different dataset IDs in a single test?

File: src/dataset-builder/ConceptSetCreator.test.ts [hunk 6]
```diff
@@ -1,22 +1,22 @@
...
+  const datasetDetails = dummyDatasetDetails(datasetId);
 
   const renderConceptSetCreator = () => {
     const conceptSetUpdater = jest.fn();
     const onStateChange = jest.fn();
     render(
       h(ConceptSetCreator, {
-        snapshotBuilderSettings: datasetDetails.snapshotBuilderSettings as SnapshotBuilderSettings,
+        dataset: datasetDetails,
```
Comment (pshapiro4broad) [inline src/dataset-builder/ConceptSetCreator.test.ts]: Usually I like to declare the local variable with the same name as the property argument so this can be written as:
```suggestion
        dataset,
```

Maybe part of the confusion is that `dummyDatasetDetails` returns `DatasetModel` instead of a "dataset details" object so its name should be `dummyDatasetModel`, so it wouldn't look odd to assign its result to a local variable called `dataset`.

File: src/dataset-builder/DatasetBuilder.ts [hunk 7]
```diff
@@ -709,12 +709,12 @@ export const DatasetBuilderView: React.FC<DatasetBuilderProps> = (props) => {
...
                   : div(['No Dataset Builder Settings Found']);
               case 'domain-criteria-selector':
-                return h(DomainCriteriaSelector, { state: datasetBuilderState, onStateChange });
+                return h(DomainCriteriaSelector, { state: datasetBuilderState, onStateChange, datasetId });
               case 'concept-set-creator':
                 return datasetDetails.state.snapshotBuilderSettings
                   ? h(ConceptSetCreator, {
                       onStateChange,
-                      snapshotBuilderSettings: datasetDetails.state.snapshotBuilderSettings,
+                      dataset: datasetDetails.state,
```
Comment (pshapiro4broad) [inline src/dataset-builder/DatasetBuilder.ts]: Here's another case where our names seem to be disagreeing. We have the names `dataset` `DatasetModel` and `datasetDetails` that all seem to refer to the same thing. If they do, they should probably all have the same name instead of different ones. It's probably OK to have `DatasetModel` and then assign it to `dataset` since `Model` doesn't really mean much, but mixing in `details` makes it confusing IMO.

File: src/dataset-builder/DomainCriteriaSelector.test.ts [hunk 8]
```diff
@@ -32,15 +33,15 @@ describe('DomainCriteriaSelector', () => {
 
   it('renders the domain criteria selector', async () => {
     // Arrange
-    render(h(DomainCriteriaSelector, { state, onStateChange: jest.fn() }));
+    render(h(DomainCriteriaSelector, { state, onStateChange: jest.fn(), datasetId }));
```
Comment (pshapiro4broad) [inline src/dataset-builder/DomainCriteriaSelector.test.ts:36]: This test now passes in a value for `datasetId`. Is this state validated in the test? That is, does/can the test somehow assert that the `datasetId` was stored and used as expected?

File: src/dataset-builder/TestConstants.ts [hunk 9]
```diff
@@ -100,3 +100,31 @@ export const dummyDatasetDetails = (datasetId: string): DatasetModel => ({
...
+  { id: 203, name: 'Heart Surgery', count: 100, hasChildren: false },
+  { id: 204, name: 'Cancer Surgery', count: 100, hasChildren: false },
+
+  { id: 300, name: 'Observation', count: 100, hasChildren: true },
+  { id: 301, name: 'Blood Pressure', count: 100, hasChildren: false },
+  { id: 302, name: 'Weight', count: 100, hasChildren: false },
+  { id: 303, name: 'Height', count: 100, hasChildren: false },
+];
+
+export const mockGetConceptForId = (id: number): Concept => {
```
Comment (pshapiro4broad) [inline src/dataset-builder/TestConstants.ts]: The distinction is fuzzy but for me `dummy` is different than `mock`. IMO mock is a method that's stubbed as a no-op and is used for verification in a test. Whereas a dummy is a fake value used for testing. I guess we do talk about using "mocked" APIs for our stubbed ajax calls.. The main danger of using `mock` in this name is that a reader might assume that it means something mocked using Jest vs a method returning dummy data.

**Review by pshapiro4broad — APPROVED** [review]
Looks good overall but I do have a few comments to address / reply to.

File: src/dataset-builder/ConceptSelector.ts (hunk 4 above)
Comment (rjohanek) [reply src/dataset-builder/ConceptSelector.ts]: Yeah I think it makes sense just to store the whole concept and remove the getConcept call

File: src/dataset-builder/ConceptSetCreator.test.ts (hunk 5 above)
Comment (rjohanek) [reply src/dataset-builder/ConceptSetCreator.test.ts]: yeah I don't think it needs an id parameter either
